### PR TITLE
Sharing ephemeral streams of location data

### PR DESCRIPTION
**Pull request by stefanceriu**

### CODE REVIEW

File: proposals/3672-ephemeral-location-streaming.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# MSC3672 - Sharing ephemeral streams of location data
+
+## Problem
+
+[MSC3489](https://github.com/matrix-org/matrix-doc/blob/matthew/location-streaming/proposals/3489-location-streaming.md) 
+focuses on streaming persistent location data for applications that require
+historical knowledge. 
+
+We also need the ability to share this data in a non-persistent way for cases in 
+which privacy is a concern, like user locations.
```
Comment (deepbluev7) [inline proposals/3672-ephemeral-location-streaming.md:10]: I think this needs a bit of a longer justification of why this is needed and why EDUs are the way to go. Some ideas:

- EDUs are not persisted in the timeline -> A user can only access the ones they received originally.
- EDUs can be stored indefinitely on the server, but because they are E2EE, this should not cause issues.
- What are the semantics for the last EDU? When do they expire? Do you always get the last sent one on initial sync? Does a client need to clobber that with a dummy?

File: proposals/3672-ephemeral-location-streaming.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+The Megolm keys required to decrypt this EDU should be shared over Olm just like
+regular encrypted timeline events.
+
+## Alternatives
+
+Alternatively, we could negotiate a WebRTC data channel using [MSC3401](https://github.com/matrix-org/matrix-doc/blob/matthew/group-voip/proposals/3401-group-voip.md) and stream low-latency geospatial data over the 
+participating devices in the room. However it would be useful to support plain 
+HTTP like the rest of Matrix and requiring a WebRTC stack is prohibitively 
+complicated for simple clients (e.g. basic IOT devices reporting spatial telemetry).
+
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply proposals/3672-ephemeral-location-streaming.md:88]: Thank you, added a paragraph about it.

File: proposals/3672-ephemeral-location-streaming.md (hunk 1 above)
Comment (stefanceriu) [reply proposals/3672-ephemeral-location-streaming.md]: Thank you. I've added more details for the first two points but perhaps the last one should be taken up in [MSC2477](https://github.com/matrix-org/matrix-doc/pull/2477)? 
I don't think there should be any special behavior for location sharing.

Comment (deepbluev7) [reply proposals/3672-ephemeral-location-streaming.md]: It might make sense to pick that up in the justification for a bit, i.e. "You can make EDUs expire after some time, so they never get delivered to devices, which are offline for some time" or so, if that is actually possible with 2477. It's been a while since I read it.

File: proposals/3672-ephemeral-location-streaming.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+are currently used for typing notifications, event receipts, and presence 
+updates. As an extra precaution they can also be encrypted as defined in [MSC3673](https://github.com/matrix-org/matrix-doc/pull/3673).
+
+We will start by introducing a new boolean property on `m.beacon_info` called 
+`live` which will mark the start of an user's intent to share ephemeral location 
+information.
+
+```json5
+{
+    "type": "m.beacon_info.@stefan:matrix.org",
```
Comment (andybalaam) [inline proposals/3672-ephemeral-location-streaming.md:36]: I suggest this should be called `m.beacon.*` - in my thinking this single event is a beacon, and each location event is a pulse of light from that beacon.

Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md:36]: And so the content property would become `m.beacon` instead of `m.beacon_info`

File: proposals/3672-ephemeral-location-streaming.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    "state_key": "@stefan:matrix.org",
```
Comment (andybalaam) [inline proposals/3672-ephemeral-location-streaming.md:37]: The fact that this is a state event has some pros and cons.

**Cons:**
* If someone ever shares their location in a room, this fact (but not the actual location) will persist indefinitely and unencrypted in the room state.
    * "persist indefinitely" will become untrue when we get deletable state. (I can't find the MSC for this, but I heard there was one)
    * "unencrypted" will become untrue when we get encrypted state [MSC3414](https://github.com/matrix-org/matrix-spec-proposals/pull/3414)
* If we want multiple beacons for a single user, we need to hack the event type to contain a counter or similar, because the combination of event type and state key makes a unique key in the room state, and the state key must be exactly the user's mxid to prevent others overwriting it.
    * The need to do this will go away if encrypted state [MSC3414](https://github.com/matrix-org/matrix-spec-proposals/pull/3414) allows us to use any state key, instead of having to use exactly our mxid.
* If location events are arriving frequently, clients can see the ID of the original event referred to there, so can explicitly fetch it when needed.
    * However, for some types of location sharing e.g. tracking a parcel delivery, the location updates will be infrequent. (Is this type of use case relevant to ephemeral location sharing though, or only [MSC3489](https://github.com/matrix-org/matrix-spec-proposals/blob/matthew/location-streaming/proposals/3489-location-streaming.md) (live location with history) ?)

**Pros:**
* Arguably, the fact that someone is sharing a location is immediately relevant when you enter a room
    * But without a location event, the beacon itself is arguably useless, and since we are emitting EDUs, we probably can't expect to go and fetch them if we missed them.

File: proposals/3672-ephemeral-location-streaming.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    "content": {
+        "m.beacon_info": {
+            "description": "Stefan's live location",
+            "timeout": 600000, // how long from the last event until we consider the beacon inactive in milliseconds
+            "live": true // this is a live location beacon
+        },
+        "m.ts": 1436829458432, // creation timestamp of the beacon on the client
+        "m.asset": {
+            "type": "m.self.live" // live user location tracking
```
Comment (andybalaam) [inline proposals/3672-ephemeral-location-streaming.md:46]: Should this be `m.self`, matching [MSC3488 location sharing](https://github.com/matrix-org/matrix-spec-proposals/blob/matthew/location/proposals/3488-location.md)? How does it help us to make it different?

File: proposals/3672-ephemeral-location-streaming.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+Multiple live beacons on the same timeline have the option of either aggregating
+all available location EDUs or just the ones referencing a particular 
+`beacon_info`.
+
+Subsequently clients will start sending beacon data EDUs to the new 
+`rooms/{roomId}/ephemeral/{eventType}/{txnId}` endpoint where `eventType` equals 
+`m.beacon` with the same location payload as defined in [MSC3489](https://github.com/matrix-org/matrix-doc/pull/3489).
+
+
+```json5
```
Comment (andybalaam) [inline proposals/3672-ephemeral-location-streaming.md:64]: Is there a reason why this example is missing the outer event info like `type`, `sender`? (Perhaps because this is an EDU?) If not, it would help me if it were included.

File: proposals/3672-ephemeral-location-streaming.md (hunk 3 above)
Comment (andybalaam) [inline proposals/3672-ephemeral-location-streaming.md:36]: This example type `m.beacon_info.@stefan:matrix.org` is not helpful: the postfix to the type is designed to allow us to have multiple state events for a user all with the same state key (because setting the state key to the user's mxid is the only way to prevent other people overwriting our state).

So, the type in this example should be e.g. `m.beacon_info.beacon_12345`. See this [comment on MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489#discussion_r823644864).

**Review by andybalaam — CHANGES_REQUESTED** [review]

**Comment by anoadragon453** [comment]
Heads up that @stefanceriu has moved on to other work. In their place I'll be acting as a [shepherd](https://spec.matrix.org/unstable/proposals/#process) and will be updating this MSC as new feedback comes in.

File: proposals/3672-ephemeral-location-streaming.md (hunk 4 above)
Comment (kerryarchibald) [reply proposals/3672-ephemeral-location-streaming.md:37]: Pros: A state event allows us to resume emitting EDUs when restarting the client with a location sharing session.

File: proposals/3672-ephemeral-location-streaming.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+{
+    "m.relates_to": { // from MSC2674: https://github.com/matrix-org/matrix-doc/pull/2674
+        "rel_type": "m.reference", // from MSC3267: https://github.com/matrix-org/matrix-doc/pull/3267
```
Comment (deepbluev7) [inline proposals/3672-ephemeral-location-streaming.md:67]: Could we lock down what `m.reference` is supposed to mean first? Some people say it is supposed to be a replacement for replies, others say it is for "other" references.

File: proposals/3672-ephemeral-location-streaming.md (hunk 3 above)
Comment (deepbluev7) [inline proposals/3672-ephemeral-location-streaming.md:36]: Since I commented this on a different MSC about location sharing, but this one does the same thing: https://github.com/matrix-org/matrix-spec-proposals/pull/3489#discussion_r781654176

Please don't use arbitrary event types. It makes state events very hard to parse for some clients (it was a complaint about account data types for SSSS, but same applies here) and even worse, how are you going to use a power_level to restrict sending such events?

File: proposals/3672-ephemeral-location-streaming.md (hunk 4 above)
Comment (ara4n) [reply proposals/3672-ephemeral-location-streaming.md:37]: Pros: A state event means that if a beacon is intermittently emitting EDUs (because it's coming in & out of connectivity, or just updates infrequently) you still immediately know semantically that there's a beacon out there which is trying to share its location, which could be super valuable data (especially if in practice they are in dire straits and currently failing to emit EDUs; it could basically act as an SOS. I've actually seen something like this happen IRL, depressingly).

I feel fairly strongly that the fact that there is a beacon in the room emitting location data belongs in room state - just like if there was a call ongoing in the room, or if a given user is present in the room, etc.  However, agreed that from a privacy perspective, we should actually delete old state events (https://github.com/matrix-org/matrix-spec/issues/43) and encrypt them too (MSC3414)

File: proposals/3672-ephemeral-location-streaming.md (hunk 3 above)
Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md:36]: I also find arbitrary event types quite troubling. If [MSC3414](https://github.com/matrix-org/matrix-spec-proposals/pull/3414) allows us to avoid this, it would be great.

Comment (turt2live) [reply proposals/3672-ephemeral-location-streaming.md:36]: This is jarring in code review for beacons on element-web. Please let's make events have a single canonical type with variable information either inside the content or refactored into another data structure.

File: proposals/3672-ephemeral-location-streaming.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+of its own as they're 1:1, single message per transaction and not intended for 
+conversational data.
+
+## Unstable prefix
+
+ * `m.beacon_info.*` should be referred to as `org.matrix.msc3489.beacon_info.*` until this MSC lands.
+ * `m.beacon` should be referred to as `org.matrix.msc3489.beacon` until this MSC lands.
+ * `m.location` should be referred to as `org.matrix.msc3488.location.*` until MSC3488 lands.
+ * `m.ts` should be referred to as `org.matrix.msc3488.ts.*` until MSC3488 lands.
+ * `m.asset` should be referred to as `org.matrix.msc3488.asset.*` until MSC3488 lands.
```
Comment (bmarty) [inline proposals/3672-ephemeral-location-streaming.md:108]: I think there are 3 typos on these 3 lines. I do not understand why we have `.*` suffixes.

Comment (bmarty) [reply proposals/3672-ephemeral-location-streaming.md:108]: I see, but does it apply to `m.location`, `m.ts` and `m.asset`?

This:

image

Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md:108]: Oh I'm sorry you're absolutely right, thanks!

File: proposals/3672-ephemeral-location-streaming.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    "m.ts": 1636829458432,
+}
+```
+
+These will reach clients through `/sync`s `ephemeral` dictionary with the same 
+payload but with the addition of a `sender` which the clients can aggregate user 
+locations on.
+
+When the user decides they would like to stop sharing their live location the 
+original `m.beacon_info`'s `live` property should be set to `false`.
```
Comment (anoadragon453) [inline proposals/3672-ephemeral-location-streaming.md:83]: I'm not sure I like the idea of the `live` field being used to both indicate:

1. Whether ephemeral location updates are intended to be used with it.
2. Whether the location sharing session is currently active.

Mainly, I find it be a bit of a bad smell that `"live": false` and no "live" field mean different things.

One possible alternative is to instead indicate the end of a live location sharing session by updating the state event to have a `timeout` value of 0. This would immediately end the session, and would replace the need to set `live` to false.

`live` could then continue to be used to indicate the difference between an ephemeral vs. historical location sharing session; which is still an important property, i.e. if clients wanted to render historical location sharing sessions with a "location history will be kept" warning or what have you.

I might also advocate for renaming `live`, as I would still consider a **historical** location sharing session "live" while it's running. Perhaps `"persistent"`?

Comment (Johennes) [reply proposals/3672-ephemeral-location-streaming.md:83]: Or maybe an enum-like `type = ephemeral | persistent | ...` to allow for later addition of other location sharing styles without having to evaluate several booleans?

File: proposals/3672-ephemeral-location-streaming.md (hunk 7 above)
Comment (turt2live) [reply proposals/3672-ephemeral-location-streaming.md:67]: for clarity: let's not argue about it here. This is https://github.com/matrix-org/matrix-spec-proposals/pull/3267

File: proposals/3672-ephemeral-location-streaming.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+```json5
+{
+    "type": "m.beacon_info",
+    "state_key": "@stefan:matrix.org",
+    "content": {
+        "m.beacon_info": {
+            "timeout": 600000,
+            "description": "Stefan's live location",
+            "live": true
+        },
```
Comment (anoadragon453) [inline proposals/3672-ephemeral-location-streaming.md:72]: Why do we arbitrarily wrap these fields in another `m.beacon_info` dictionary, when the event type already marks this state event as `m.beacon_info`?

File: proposals/3672-ephemeral-location-streaming.md (hunk 9 above)
Comment (anoadragon453) [reply proposals/3672-ephemeral-location-streaming.md:118]: That, while still including a way for clients to update the `m.beacon_info` state event to indicate that the session has ended, yes.

File: proposals/3672-ephemeral-location-streaming.md (hunk 10 above)
Comment (anoadragon453) [reply proposals/3672-ephemeral-location-streaming.md:72]: As far as I can tell, and after asking around, there isn't a reason for this. So I'm going to remove it.

Comment (dkasak) [reply proposals/3672-ephemeral-location-streaming.md]: Probably foreshadowing [extensible events](https://github.com/matrix-org/matrix-spec-proposals/blob/matthew/msc1767/proposals/1767-extensible-events.md)?

File: proposals/3672-ephemeral-location-streaming.md [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+The `state_key` of this state event must be send to the sender's MXID. As per [event auth rules](https://spec.matrix.org/v1.2/rooms/v9/#authorization-rules), this restricts the state event to only be editable by the sender.
+
+**`m.beacon_info` state event `content` field definitions**
+
+| key | type | value | description | Required |
+| ---- | ----| ----- | ----------- | -------- |
+| `timeout` | int | A positive number of milliseconds | The maximum length of the location sharing session, relative to `m.ts`. | yes
+| `description` | string | Optional descriptive text | A human-readable description of the live location sharing session. | no |
+| `live` | bool | true | A boolean describing whether the location sharing session is currently active. Also denotes this session as ephemeral. | yes
+| `m.ts` | int | [Unix timestamp](https://en.wikipedia.org/wiki/Unix_time) | The timestamp of when the location sharing session was started by the sender. | yes
```
Comment (kerryarchibald) [inline proposals/3672-ephemeral-location-streaming.md:41]: Can we please add a clarifying note that this is ms since the Unix epoch? the wiki link refers to unix timestamps in seconds since epoch.

File: proposals/3672-ephemeral-location-streaming.md (hunk 3 above)
Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md]: I've become convinced that the current naming is fine - a beacon_info is a collection of beacons.

File: proposals/3672-ephemeral-location-streaming.md (hunk 4 above)
Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md:60]: I'm happy with this: state events are for representing room state i.e. things that are true, like the fact that someone is sharing their location.

File: proposals/3672-ephemeral-location-streaming.md [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+We define this new property in order to allow clients to distinguish between 
+share types without potentially overwriting static ones.
+
+Multiple live beacons on the same timeline have the option of either aggregating
+all available location EDUs or just the ones referencing a particular 
+`beacon_info`.
+
+Subsequently clients will start sending beacon data EDUs to the new 
+`rooms/{roomId}/ephemeral/{eventType}/{txnId}` endpoint where `eventType` equals 
+`m.beacon` with the same location payload as defined in [MSC3489](https://github.com/matrix-org/matrix-doc/pull/3489).
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md]: I'm ok with `m.beacon`

File: proposals/3672-ephemeral-location-streaming.md (hunk 3 above)
Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md]: Soon, this MSC will be updated to avoid variable types, like MSC3489.

Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md]: Soon, this MSC will be updated to avoid variable event types, like MSC3489

Comment (anoadragon453) [reply proposals/3672-ephemeral-location-streaming.md]: (It already has been btw.)

File: proposals/3672-ephemeral-location-streaming.md (hunk 10 above)
Comment (anoadragon453) [reply proposals/3672-ephemeral-location-streaming.md]: Ah, good point! Perhaps we should add this back in then.

Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md]: There was some discussion while we were implementing MSC3489 and it was felt that adding a `m.beacon_info` section might slightly miss the point of extensible events. Objects like this are intended to be reusable chunks of code, so while e.g. `m.location` makes sense because it will be re-used in both `m.location` and `m.beacon` events, probably it doesn't make sense here.

Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md]: I've remodelled this MSC so it leans heavily on MSC3489, which has removed this `m.beacon_info` structure.

**Review by andybalaam — APPROVED** [review]

File: proposals/3672-ephemeral-location-streaming.md (hunk 1 above)
Comment (deepbluev7) [reply proposals/3672-ephemeral-location-streaming.md]: Took me a while to find, but yes, that looks reasonable. Does that work when the EDU is encrypted? Because then the event type is m.room.encrypted, isn't it?

Comment (andybalaam) [reply proposals/3672-ephemeral-location-streaming.md]: Sorry, I should have provided a proper link.

I expect that if we later add the ability to specify delivery guarantees for EDUs, we will include that in the unencrypted part. Before we have that, servers will have to apply the same policy to all user-defined encrypted EDUs.

File: proposals/3672-ephemeral-location-streaming.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+suggest 30 minutes would be a suitable timeout.
+
+## Security considerations
+
+The security considerations are identical to MSC3489, except that EDUs
+provide a better privacy situation because they are not persisted long-term.
+
+## Alternatives
+
+Alternatively, we could negotiate a WebRTC data channel using
```
Comment (Half-Shot) [inline proposals/3672-ephemeral-location-streaming.md:65]: The upcoming MatrixRTC MSC may help here. It would allow you to negotiate other RTC mechanisms, perhaps something as simple as a plain websocket-style communication mechanism.
